### PR TITLE
GO-3286: fix not working remove and delete notifications from member side

### PR DESCRIPTION
**Pull request by AnastasiaShemyakinskaya**
https://linear.app/anytype/issue/GO-3286/fix-not-working-remove-and-delete-notifications-from-member-side

## Summary by CodeRabbit

- **New Features**
	- Enhanced security and stability in notification handling through improved synchronization.
	- Updated account status management to reflect current data accuracy.

- **Bug Fixes**
	- Corrected account status parameters in testing scenarios to align with actual application behavior.

- **Tests**
	- Added comprehensive tests for new notification functionalities and account status updates.

- **Documentation**
	- Introduced interfaces for better modularity and testing in notification and identity services.

### CODE REVIEW

File: space/internal/components/aclnotifications/notifications.go [hunk 1]
```diff
@@ -71,9 +75,13 @@ func (n *aclNotificationSender) Run(ctx context.Context) (err error) {
 }
 
 func (n *aclNotificationSender) Close(ctx context.Context) (err error) {
-	if err := n.batcher.Close(); err != nil {
-		logger.Errorf("failed to close batcher, %s", err)
-	}
+	go func() {
+		n.mx.Lock()
```
Comment (fat-fellow) [inline space/internal/components/aclnotifications/notifications.go:79]: batcher is thread safe

File: space/internal/components/aclnotifications/notifications.go [hunk 2]
```diff
@@ -71,9 +75,13 @@ func (n *aclNotificationSender) Run(ctx context.Context) (err error) {
 }
 
 func (n *aclNotificationSender) Close(ctx context.Context) (err error) {
-	if err := n.batcher.Close(); err != nil {
```
Comment (mcrakhman) [inline space/internal/components/aclnotifications/notifications.go:74]: Why was this needed? It looks weird, we should wait when batcher closes synchronously

File: space/internal/components/aclobjectmanager/aclobjectmanager.go [hunk 3]
```diff
@@ -157,14 +157,19 @@ func (a *aclObjectManager) process() {
 
 func (a *aclObjectManager) processAcl() (err error) {
 	var (
-		common   = a.sp.CommonSpace()
-		acl      = common.Acl()
-		aclState = acl.AclState()
+		common         = a.sp.CommonSpace()
```
Comment (mcrakhman) [inline space/internal/components/aclobjectmanager/aclobjectmanager.go:160]: Tbh looking at this now, I would've done it like this:
- remove checking for self status from process states, to make it simpler
- create another method `isAccountRemoved` which would be very simple: call `aclState.Permissions(aclState.Identity())`check if the account was removed (you can also add there the logic with upToDate)

File: space/internal/components/aclnotifications/notifications.go (hunk 1 above)
Comment (AnastasiaShemyakinskaya) [reply space/internal/components/aclnotifications/notifications.go:79]: It doesn't relate to badger, notification can be send in two different processes (when space `loader` is inited and in `UpdateAcl`) in acl object manager. Space loader can close components like acl notification service, while another process (`UpdateAcl`) is still executed. So we avoid racing here

File: space/internal/components/aclnotifications/notifications.go [hunk 4]
```diff
@@ -71,6 +75,8 @@ func (n *aclNotificationSender) Run(ctx context.Context) (err error) {
 }
 
 func (n *aclNotificationSender) Close(ctx context.Context) (err error) {
+	n.mx.Lock()
```
Comment (mcrakhman) [inline space/internal/components/aclnotifications/notifications.go:78]: I guess we don't need a mutex here

Comment (fat-fellow) [reply space/internal/components/aclnotifications/notifications.go:78]: I've said that before also

Comment (mcrakhman) [reply space/internal/components/aclnotifications/notifications.go:78]: We already revised approach a little bit, so some changes will follow

**Review by mcrakhman — APPROVED** [review]

**Pull request merged**